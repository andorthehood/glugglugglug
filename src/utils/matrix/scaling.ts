/**
 * Creates a matrix that scales along each axis independently.
 * @param sx Scale factor along the x axis.
 * @param sy Scale factor along the y axis.
 * @param sz Scale factor along the z axis.
 * @param dst Destination matrix to write into. When omitted a new matrix is created.
 * @returns A scaling matrix.
 */
export function scaling(
	sx: number,
	sy: number,
	sz: number,
	dst: number[] = new Array(16),
) {
	dst[0] = sx;
	dst[1] = 0;
	dst[2] = 0;
	dst[3] = 0;
	dst[4] = 0;
	dst[5] = sy;
	dst[6] = 0;
	dst[7] = 0;
	dst[8] = 0;
	dst[9] = 0;
	dst[10] = sz;
	dst[11] = 0;
	dst[12] = 0;
	dst[13] = 0;
	dst[14] = 0;
	dst[15] = 1;
	return dst;
}
