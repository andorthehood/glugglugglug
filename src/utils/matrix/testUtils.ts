export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
