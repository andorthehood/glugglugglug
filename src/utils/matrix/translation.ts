/**
 * Creates a 4x4 matrix that translates by the supplied offsets.
 * @param tx Translation along the x axis.
 * @param ty Translation along the y axis.
 * @param tz Translation along the z axis.
 * @param dst Destination matrix to write into. When omitted a new matrix is created.
 * @returns A translation matrix.
 */
export function translation(
	tx: number,
	ty: number,
	tz: number,
	dst: number[] = new Array(16),
) {
	dst[0] = 1;
	dst[1] = 0;
	dst[2] = 0;
	dst[3] = 0;
	dst[4] = 0;
	dst[5] = 1;
	dst[6] = 0;
	dst[7] = 0;
	dst[8] = 0;
	dst[9] = 0;
	dst[10] = 1;
	dst[11] = 0;
	dst[12] = tx;
	dst[13] = ty;
	dst[14] = tz;
	dst[15] = 1;
	return dst;
}
