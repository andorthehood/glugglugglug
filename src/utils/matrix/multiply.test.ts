import { describe, expect, it } from "vitest";
import { multiply } from "./multiply";
import { scaling } from "./scaling";
import { translation } from "./translation";
import { formatMatrix } from "./testUtils";

describe("multiply", () => {
	it("multiplies two matrices in column-major order", () => {
		const a = translation(1, 2, 3);
		const b = scaling(2, 3, 4);
		const productAB = multiply(a, b);

		expect(productAB).toBe(a);
		expect(formatMatrix(productAB)).toMatchSnapshot();

		const c = scaling(2, 3, 4);
		const d = translation(1, 2, 3);
		const productBA = multiply(c, d);

		expect(productBA).toBe(c);
		expect(formatMatrix(productBA)).toMatchSnapshot();
	});

	it("writes into a provided destination without mutating inputs", () => {
		const a = translation(1, 2, 3);
		const b = scaling(2, 3, 4);
		const dst = new Array<number>(16);
		const result = multiply(a, b, dst);

		expect(result).toBe(dst);
		expect(formatMatrix(result)).toMatchSnapshot();
		expect(formatMatrix(a)).toEqual(formatMatrix(translation(1, 2, 3)));
	});
});
