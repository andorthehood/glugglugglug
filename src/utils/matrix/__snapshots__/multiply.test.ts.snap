// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`multiply > multiplies two matrices in column-major order 1`] = `
[
  [
    2,
    0,
    0,
    0,
  ],
  [
    0,
    3,
    0,
    0,
  ],
  [
    0,
    0,
    4,
    0,
  ],
  [
    1,
    2,
    3,
    1,
  ],
]
`;

exports[`multiply > multiplies two matrices in column-major order 2`] = `
[
  [
    2,
    0,
    0,
    0,
  ],
  [
    0,
    3,
    0,
    0,
  ],
  [
    0,
    0,
    4,
    0,
  ],
  [
    2,
    6,
    12,
    1,
  ],
]
`;

exports[`multiply > writes into a provided destination without mutating inputs 1`] = `
[
  [
    2,
    0,
    0,
    0,
  ],
  [
    0,
    3,
    0,
    0,
  ],
  [
    0,
    0,
    4,
    0,
  ],
  [
    1,
    2,
    3,
    1,
  ],
]
`;
